routers/api/v1/activitypub/person.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package activitypub
+
+import (
+	"net/http"
+	"strings"
+
+	"code.gitea.io/gitea/modules/activitypub"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	ap "github.com/go-ap/activitypub"
+)
+
+// Person function returns the Person actor for a user
+func Person(ctx *context.APIContext) {
+	// swagger:operation GET /activitypub/user/{username} activitypub activitypubPerson
+	// ---
+	// summary: Returns the Person actor for a user
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: username
+	//   in: path
+	//   description: username of the user
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/ActivityPub"
+
+	link := strings.TrimSuffix(setting.AppURL, "/") + "/api/v1/activitypub/user/" + ctx.ContextUser.Name
+	person := ap.PersonNew(ap.IRI(link))
+
+	person.Name = ap.NaturalLanguageValuesNew()
+	err := person.Name.Set("en", ap.Content(ctx.ContextUser.FullName))
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "Set Name", err)
+		return
+	}
+
+	person.PreferredUsername = ap.NaturalLanguageValuesNew()
+	err = person.PreferredUsername.Set("en", ap.Content(ctx.ContextUser.Name))
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "Set PreferredUsername", err)
+		return
+	}
+
+	person.URL = ap.IRI(ctx.ContextUser.HTMLURL())
+
+	person.Icon = ap.Image{
+		Type:      ap.ImageType,
+		MediaType: "image/png",
+		URL:       ap.IRI(ctx.ContextUser.AvatarLink()),
+	}
+
+	person.Inbox = ap.IRI(link + "/inbox")
+	person.Outbox = ap.IRI(link + "/outbox")
+
+	person.PublicKey.ID = ap.IRI(link + "#main-key")
+	person.PublicKey.Owner = ap.IRI(link)
+
+	publicKeyPem, err := activitypub.GetPublicKey(ctx.ContextUser)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetPublicKey", err)
+		return
+	}
+	person.PublicKey.PublicKeyPem = publicKeyPem
+
+	binary, err := person.MarshalJSON()
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "Serialize", err)
+		return
+	}
+
+	var jsonmap map[string]interface{}
+	err = json.Unmarshal(binary, &jsonmap)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "Unmarshal", err)
+		return
+	}
+
+	jsonmap["@context"] = []string{"https://www.w3.org/ns/activitystreams", "https://w3id.org/security/v1"}
```

Comment:
I still think the json conversion is really bad. I hoped a simple workaround could be

```go
type PersonContext struct {
	AtContext []ap.IRI `json:"@context"`
	*ap.Person
}

pc := &PersonContext{
	ap.JsonLDContext,
	person,
}

json.Marshal(pc)
```
This works for every type but not for `*ap.Person`, I don't know why.
Otherwise see the comment from @mariusor

Reply:
The objects in the activitypub package should be marshaled using the `go-ap/jsonld` package which makes usage of the custom tags for the properties. It however should not behave differently for `Person`. If it does, I'd appreciate a repro case on the mailing list: https://lists.sr.ht/~mariusor/activitypub-go and I'll fix it ASAP.

Reply:
I don't know what's the problem at the moment. Here is my test code:

```go
type Dummy struct {
	Id string
}

type DummyContext struct {
	AtContext []ap.IRI `json:"@context"`
	*Dummy
}

dc := &DummyContext{
	ap.JsonLDContext,
	&Dummy{Id:"test"},
}

x, _ := json.Marshal(dc)
log.Error("%#v", dc) // &activitypub.DummyContext{AtContext:[]activitypub.IRI{https://www.w3.org/ns/activitystreams, https://w3id.org/security/v1}, Dummy:(*activitypub.Dummy)(0xc0051b3c90)}
log.Error("%s", string(x)) // {"@context":["https://www.w3.org/ns/activitystreams","https://w3id.org/security/v1"],"Id":"test"}

type PersonContext struct {
	AtContext []ap.IRI `json:"@context"`
	*ap.Person
}

pc := &PersonContext{
	ap.JsonLDContext,
	person,
}

x, _ = json.Marshal(pc)
log.Error("%#v", pc) // activitypub.Actor[Person] { } // Where is my wrapper type?!
log.Error("%s", string(x)) // {"id":"...","type":"Person",...}} // missing @context
```

Have a look at the comments in the `log.Error` lines.